### PR TITLE
Tweak wording when encountering `fn` call in pattern

### LINKED ISSUE #60642: Provide more of an explanation when using function as pattern
_(Follow up to #60635 and https://github.com/rust-lang/rust/pull/60641)_

[When trying to match on the result of a `fn` call](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=046b07a1f14bab1e3b7d3b1708f30dd8), we emit:

```
error[E0164]: expected tuple struct/variant, found method `<A>::new`
  --> src/lib.rs:15:9
   |
15 |         A::new() => (),
   |         ^^^^^^^^ not a tuple variant or struct
```

We should explain that `fn` calls are not allowed ever in patterns, only refutable (or even irrefutable, as you can match on a struct, although it is silly) types, possibly [linking to (or cribbing text from) the appropriate chapter in The Book](https://doc.rust-lang.org/book/ch18-00-patterns.html?highlight=pattern#patterns-and-matching).

### PR DESCRIPTION
**Pull request by estebank**
Fix #60642

### CODE REVIEW

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -1080,8 +1080,18 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
...
                               res.descr(),
                               hir::print::to_string(tcx.hir(), |s| s.print_qpath(qpath, false)));
-            struct_span_err!(tcx.sess, pat.span, E0164, "{}", msg)
-                .span_label(pat.span, "not a tuple variant or struct").emit();
+            let mut err = struct_span_err!(tcx.sess, pat.span, E0164, "{}", msg);
+            match (res, &pat.node) {
+                (Res::Def(DefKind::Fn, _), _) | (Res::Def(DefKind::Method, _), _) => {
+                    err.span_label(pat.span, "`fn` calls are not allowed in patterns");
+                    err.help("for more information, visit https://doc.rust-lang.org/book/\
+                              ch18-00-patterns.html?highlight=pattern#patterns-and-matching");
```
Comment (varkor) [inline src/librustc_typeck/check/_match.rs]: Was the highlight intentional?

Comment (estebank) [reply src/librustc_typeck/check/_match.rs]: nope

Comment (estebank) [reply src/librustc_typeck/check/_match.rs]: fixed

Comment (varkor) [reply src/librustc_typeck/check/_match.rs]: (It's outdated in the `.stderr`s.)

Comment (estebank) [reply src/librustc_typeck/check/_match.rs]: oops, fixed

**Pull request merged**